src/Xrefcheck/Verify.hs

```diff
@@ -231,22 +274,39 @@ data NeedsCaching key
 -- then the @action@ will not be executed, and the value is added to the accumulator list.
 -- After the whole list has been traversed, the accumulator is traversed once again to ensure
 -- every asynchronous action is completed.
+-- If interrupted by AsyncException, returns this exception and list of already calcualted results
+-- (their subset can be arbitrary). Computations that were not ended till this moment are cancelled.
 forConcurrentlyCaching
   :: Ord cacheKey
-  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO [b]
-forConcurrentlyCaching list needsCaching action = go [] M.empty list
+  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO (Either (AsyncException, [b]) [b])
+forConcurrentlyCaching list needsCaching action = do
+  res <- go [] M.empty list
+  case res of
+    Right r -> return $ Right $ map (fromMaybe (error "impossible")) r
+      -- impossible, since no actions were cancelled
+
+    Left (exception, partialResult) -> return $  Left (exception, catMaybes partialResult)
   where
-    go acc cached (x : xs) = case needsCaching x of
-      NoCaching -> do
-        withAsync (action x) $ \b ->
-          go (b : acc) cached xs
-      CacheUnderKey cacheKey -> do
-        case M.lookup cacheKey cached of
-          Nothing -> do
-            withAsync (action x) $ \b ->
-              go (b : acc) (M.insert cacheKey b cached) xs
-          Just b -> go (b : acc) cached xs
-    go acc _ [] = for acc wait <&> reverse
+    interruptibleAction a = handle (\AsyncCancelled -> pure Nothing) (Just <$> action a)
+
+    go acc cached items = handle
```

Comment:
Hm it's a bit weird to have an exception handler in a recursive function like `go`... If the list `[a]` has 10 items, we'll end up with 10 exception handles stacked on top of each other.

I think a better place to do this would be in the `case items of [] ->` branch, where we wait for all children threads to complete:

```hs
      case items of
        [] ->
          -- Wait for all children threads to complete.
          --
          -- If, while the threads are running, the user hits Ctrl+C,
          -- a `UserInterrupt :: AsyncException` will be thrown onto the main thread.
          -- We catch it here, cancel all child threads, and return the results of only the threads that finished successfully.
          (Right . reverse <$> for acc wait)
          `catch`
          \exception -> do
            partialResults <- for acc \asyncAction -> do
              cancel asyncAction
              poll asyncAction <&> \case
                Nothing -> Nothing
                Just (Right a) -> Just a
                Just (Left _ex) -> Nothing
            pure $ Left (exception, catMaybes partialResults)
`

Reply:
in this case I was able to quickly hit Ctrl-C before items list was empty. Even when progress bar indicates some errors. This is very strange because spawning of async threads is fast. I will test this again, and if in normal scenario it works well, I agree that this looks better

Reply:
Yes, this is totaly not working for me, e.g. when I interrupt xrefheck during it checks external links on xrefcheck repo (and waiting 429's timeout with blue progress bar). Maybe we need to mask exceptions in other cases, I'll try to figure this out

Reply:
Everything except this is done

Reply:
Last  fixup uses approach from comment

Reply:
Good job, well done!

Instead of replacing `handle` from `universum` (which is actually re-exported from `safe-exceptions`) with `handle` from `exceptions`, I think it would be more idiomatic to replace it with `handleAsync` from `safe-exceptions`.

Reply:
Good job, well done!

Instead of replacing `handle` from `universum` (which is actually re-exported from `safe-exceptions`) with `handle` from `exceptions`, I think it would be more idiomatic to replace it with `handleAsync` from `safe-exceptions`.

Reply:
Also, please don't forget to add this comment near `handle`:

```hs
          -- Wait for all children threads to complete.
          --
          -- If, while the threads are running, the user hits Ctrl+C,
          -- a `UserInterrupt :: AsyncException` will be thrown onto the main thread.
          -- We catch it here, cancel all child threads,
          -- and return the results of only the threads that finished successfully.
```

Reply:
`handleAsync` is not allowing us to handle only part of Async exceptions (because there is no `handleAsyncJust`), but I don't see a problem with handling all things from `AsyncException` type. Also to avoid depending both `exceptions` and `safe-exceptions`, I changed `handleJust` here
```
      mkURIBs (URIBS.serializeURIRef' uri')
           -- Ideally, this exception should never be thrown, as the URI
           -- already *percent-encoded* with 'parseURI' from 'uri-bytestring'
           -- and 'mkURIBs' is only used to convert to 'URI' type from
           -- 'modern-uri' package.
           & handleJust (fromException @ParseExceptionBs)
           (throwError . ExternalResourceUriConversionError)
```
to one from `safe-exceptions` (it was the only use of exceptions package before)

Reply:
> `handleAsync` is not allowing us to handle only part of `Async` exceptions

That's fine, we can just re-throw the other constructors (that's what `handleJust` and `catchJust` do [under the hood anyway](https://hackage.haskell.org/package/base-4.17.0.0/docs/src/Control.Exception.Base.html#catchJust)).

So:

```hs
handleAsync
  (\case
    UserInterrupt x -> ...
    otherAsyncEx -> throwM otherAsyncEx
  )
```

> Also to avoid depending both exceptions and safe-exceptions, I changed handleJust here

Good idea!